File: Example/FWFramework/Classes/Module/Test/Toolkit/TestAdaptiveController.m
//
//  TestAdaptiveController.m
//  FWFramework_Example
//
//  Created by wuyong on 2022/8/25.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

#import "TestAdaptiveController.h"
#import "AppSwift.h"
@import FWFramework;

@interface TestAdaptiveChildController : UIViewController <FWViewController>

@property (nonatomic, assign) NSInteger index;

@end

@implementation TestAdaptiveChildController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.fw_extendedLayoutEdge = UIRectEdgeAll;
    if (self.index < 1) {
        self.fw_navigationBarStyle = FWNavigationBarStyleDefault;
    } else if (self.index < 2) {
        self.fw_navigationBarStyle = 1;
    } else if (self.index < 3) {
        self.fw_navigationBarStyle = 2;
    } else {
        self.fw_navigationBarStyle = [[@[@(-1), @(0), @(1), @(2)] fw_randomObject] integerValue];
        self.fw_navigationBarHidden = self.fw_navigationBarStyle == -1;
    }
    self.navigationItem.title = [NSString stringWithFormat:@"标题:%@ 样式:%@", @(self.index + 1), @(self.fw_navigationBarStyle)];
    
    FWWeakifySelf();
    [self fw_setRightBarItem:@"打开界面" block:^(id sender) {
        FWStrongifySelf();
        TestAdaptiveChildController *viewController = [TestAdaptiveChildController new];
        viewController.index = self.index + 1;
        [self.navigationController pushViewController:viewController animated:YES];
    }];
    [self.view fw_addTapGestureWithBlock:^(id sender) {
        FWStrongifySelf();
        TestAdaptiveChildController *viewController = [TestAdaptiveChildController new];
        viewController.index = self.index + 1;
        [self.navigationController pushViewController:viewController animated:YES];
    }];
}

@end

@interface TestAdaptiveController () <FWTableViewController>

FWPropertyWeak(UILabel *, frameLabel);
FWPropertyAssign(BOOL, hideToast);

@end

@implementation TestAdaptiveController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.fw_tabBarHidden = YES;
    [self fw_observeNotification:UIDeviceOrientationDidChangeNotification target:self action:@selector(refreshBarFrame)];
    
    if (!self.hideToast) {
        [self fw_setRightBarItem:@"启用" block:^(id sender) {
            [UINavigationController fw_enableBarTransition];
        }];
    } else {
        FWWeakifySelf();
        [self fw_setLeftBarItem:FWIcon.closeImage block:^(id  _Nonnull sender) {
            FWStrongifySelf();
            [self fw_closeViewControllerAnimated:YES];
        }];
    }
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (!self.hideToast) {
        [UIWindow fw_showMessageWithText:[NSString stringWithFormat:@"viewWillAppear:%@", @(animated)]];
    }
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    if (!self.hideToast) {
        [UIWindow fw_showMessageWithText:[NSString stringWithFormat:@"viewWillDisappear:%@", @(animated)]];
    }
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    [self refreshBarFrame];
}

- (UITableViewStyle)setupTableStyle
{
    return UITableViewStyleGrouped;
}

- (void)setupTableLayout
{
    UILabel *frameLabel = [UILabel new];
    self.frameLabel = frameLabel;
    frameLabel.numberOfLines = 0;
    frameLabel.textColor = [AppTheme textColor];
    frameLabel.font = [UIFont fw_fontOfSize:15];
    frameLabel.textAlignment = NSTextAlignmentCenter;
    [self.view addSubview:frameLabel]; {
        frameLabel.fw_layoutChain.leftWithInset(10).rightWithInset(10)
            .bottomWithInset(FWTabBarHeight + 10);
    }
    
    self.tableView.backgroundColor = [AppTheme tableColor];
    self.tableView.fw_layoutChain.horizontal().top()
        .bottomToViewTopWithOffset(self.frameLabel, -10);
}

- (void)setupSubviews
{
    [self.tableData addObjectsFromArray:@[
        @[@"状态栏切换", @"onStatusBar"],
        @[@"状态栏样式", @"onStatusStyle"],
        @[@"导航栏切换", @"onNavigationBar"],
        @[@"导航栏样式", @"onNavigationStyle"],
        @[@"标题栏颜色", @"onTitleColor"],
        @[@"大标题切换", @"onLargeTitle"],
        @[@"标签栏切换", @"onTabBar"],
        @[@"工具栏切换", @"onToolBar"],
        @[@"导航栏转场", @"onTransitionBar"],
    ]];
    if (!self.hideToast) {
        [self.tableData addObject:@[@"Present(默认)", @"onPresent"]];
        [self.tableData addObject:@[@"Present(FullScreen)", @"onPresent2"]];
        [self.tableData addObject:@[@"Present(PageSheet)", @"onPresent3"]];
        [self.tableData addObject:@[@"Present(默认带导航栏)", @"onPresent4"]];
        [self.tableData addObject:@[@"Present(Popover)", @"onPresent5:"]];
    } else {
        [self.tableData addObject:@[@"Dismiss", @"onDismiss"]];
    }
    [self.tableData addObject:@[@"设备转向", @"onOrientation"]];
}

#pragma mark - TableView

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.tableData.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [UITableViewCell fw_cellWithTableView:tableView];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    cell.textLabel.text = [rowData objectAtIndex:0];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSArray *rowData = [self.tableData objectAtIndex:indexPath.row];
    SEL selector = NSSelectorFromString([rowData objectAtIndex:1]);
    if ([self respondsToSelector:selector]) {
        FWIgnoredBegin();
        [self performSelector:selector withObject:indexPath];
        FWIgnoredEnd();
    }
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    [self refreshBarFrame];
}

#pragma mark - Protected

- (BOOL)prefersStatusBarHidden
{
    return self.fw_statusBarHidden;
}

- (UIStatusBarStyle)preferredStatusBarStyle
{
    return self.fw_statusBarStyle;
}

#pragma mark - Action

- (void)refreshBarFrame
{
    self.frameLabel.text = [NSString stringWithFormat:@"全局状态栏：%.0f 当前状态栏：%.0f\n全局导航栏：%.0f 当前导航栏：%.0f\n全局顶部栏：%.0f 当前顶部栏：%.0f\n全局标签栏：%.0f 当前标签栏：%.0f\n全局工具栏：%.0f 当前工具栏：%.0f\n全局安全区域：%@",
                            [UIScreen fw_statusBarHeight], [self fw_statusBarHeight],
                            [UIScreen fw_navigationBarHeight], [self fw_navigationBarHeight],
                            [UIScreen fw_topBarHeight], [self fw_topBarHeight],
                            [UIScreen fw_tabBarHeight], [self fw_tabBarHeight],
                            [UIScreen fw_toolBarHeight], [self fw_toolBarHeight],
                            NSStringFromUIEdgeInsets([UIScreen fw_safeAreaInsets])];
}

- (void)onStatusBar
{
    self.fw_statusBarHidden = !self.fw_statusBarHidden;
    [self refreshBarFrame];
}

- (void)onStatusStyle
{
    if (self.fw_statusBarStyle == UIStatusBarStyleDefault) {
        self.fw_statusBarStyle = UIStatusBarStyleLightContent;
    } else {
        self.fw_statusBarStyle = UIStatusBarStyleDefault;
    }
    [self refreshBarFrame];
}

- (void)onNavigationBar
{
    self.fw_navigationBarHidden = !self.fw_navigationBarHidden;
    [self refreshBarFrame];
}

- (void)onNavigationStyle
{
    if (self.fw_navigationBarStyle == FWNavigationBarStyleDefault) {
        self.fw_navigationBarStyle = 1;
    } else {
        self.fw_navigationBarStyle = FWNavigationBarStyleDefault;
    }
    [self refreshBarFrame];
}

- (void)onTitleColor
{
    self.navigationController.navigationBar.fw_titleAttributes = self.navigationController.navigationBar.fw_titleAttributes ? nil : @{NSForegroundColorAttributeName: AppTheme.buttonColor};
}

- (void)onLargeTitle
{
    self.navigationController.navigationBar.prefersLargeTitles = !self.navigationController.navigationBar.prefersLargeTitles;
    [self refreshBarFrame];
}

- (void)onTabBar
{
    self.fw_tabBarHidden = !self.fw_tabBarHidden;
    [self refreshBarFrame];
}

- (void)onToolBar
{
    if (self.fw_toolBarHidden) {
        UIBarButtonItem *item = [UIBarButtonItem fw_itemWithObject:@(UIBarButtonSystemItemCancel) target:self action:@selector(onToolBar)];
        UIBarButtonItem *item2 = [UIBarButtonItem fw_itemWithObject:@(UIBarButtonSystemItemDone) target:self action:@selector(onPresent)];
        self.toolbarItems = @[item, item2];
        self.fw_toolBarHidden = NO;
    } else {
        self.fw_toolBarHidden = YES;
    }
    [self refreshBarFrame];
}

- (void)onPresent
{
    TestAdaptiveController *viewController = [[TestAdaptiveController alloc] init];
    viewController.fw_presentationDidDismiss = ^{
        [UIWindow fw_showMessageWithText:@"fwPresentationDidDismiss"];
    };
    viewController.fw_completionHandler = ^(id  _Nullable result) {
        [UIWindow fw_showMessageWithText:@"fwCompletionHandler"];
    };
    viewController.hideToast = YES;
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)onPresent2
{
    TestAdaptiveController *viewController = [[TestAdaptiveController alloc] init];
    viewController.hideToast = YES;
    viewController.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)onPresent3
{
    TestAdaptiveController *viewController = [[TestAdaptiveController alloc] init];
    viewController.fw_presentationDidDismiss = ^{
        [UIWindow fw_showMessageWithText:@"fwPresentationDidDismiss"];
    };
    viewController.fw_completionHandler = ^(id  _Nullable result) {
        [UIWindow fw_showMessageWithText:@"fwCompletionHandler"];
    };
    viewController.hideToast = YES;
    viewController.modalPresentationStyle = UIModalPresentationPageSheet;
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)onPresent4
{
    TestAdaptiveController *viewController = [[TestAdaptiveController alloc] init];
    viewController.hideToast = YES;
    UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:viewController];
    navController.fw_presentationDidDismiss = ^{
        [UIWindow fw_showMessageWithText:@"fwPresentationDidDismiss"];
    };
    navController.fw_completionHandler = ^(id  _Nullable result) {
        [UIWindow fw_showMessageWithText:@"fwCompletionHandler"];
    };
    [self presentViewController:navController animated:YES completion:nil];
}

- (void)onPresent5:(NSIndexPath *)indexPath
{
    if (self.presentedViewController) {
        [self dismissViewControllerAnimated:YES completion:nil];
        return;
    }
    
    TestAdaptiveController *viewController = [[TestAdaptiveController alloc] init];
    viewController.hideToast = YES;
    viewController.preferredContentSize = CGSizeMake(FWScreenWidth / 2, FWScreenHeight / 2);
    [viewController fw_setPopoverPresentation:^(UIPopoverPresentationController *controller) {
        controller.barButtonItem = self.navigationItem.rightBarButtonItem;
        controller.permittedArrowDirections = UIPopoverArrowDirectionUp;
        UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
        controller.passthroughViews = [NSArray arrayWithObjects:cell, nil];
    } shouldDismiss:[@[@0, @1].fw_randomObject fw_safeBool]];
    viewController.fw_presentationDidDismiss = ^{
        [UIWindow fw_showMessageWithText:@"fwPresentationDidDismiss"];
    };
    viewController.fw_completionHandler = ^(id  _Nullable result) {
        [UIWindow fw_showMessageWithText:@"fwCompletionHandler"];
    };
    [self presentViewController:viewController animated:YES completion:nil];
}

- (void)onDismiss
{
    FWWeakifySelf();
    [self dismissViewControllerAnimated:YES completion:^{
        FWStrongifySelf();
        NSLog(@"self: %@", self);
    }];
}

- (void)onTransitionBar
{
    [self.navigationController pushViewController:[TestAdaptiveChildController new] animated:YES];
}

- (void)onOrientation
{
    if ([UIDevice fw_isDeviceLandscape]) {
        [UIDevice fw_setDeviceOrientation:UIDeviceOrientationPortrait];
    } else {
        [UIDevice fw_setDeviceOrientation:UIDeviceOrientationLandscapeLeft];
    }
    [self refreshBarFrame];
}

@end

File: Example/FWFramework/Classes/Module/Test/Toolkit/TestAdaptiveController.swift
//
//  TestAdaptiveController.swift
//  FWFramework_Example
//
//  Created by wuyong on 2022/9/20.
//  Copyright © 2022 CocoaPods. All rights reserved.
//

import FWFramework

class TestAdaptiveController: UIViewController, TableViewControllerProtocol {
    typealias TableElement = [String]

    var hideToast = false
    var globalThread = false

    private lazy var frameLabel: UILabel = {
        let result = UILabel()
        result.numberOfLines = 0
        result.textColor = AppTheme.textColor
        result.font = APP.font(15)
        result.textAlignment = .center
        return result
    }()

    func setupNavbar() {
        app.statusBarStyle = .default
        app.statusBarHidden = false
        app.observeNotification(UIDevice.orientationDidChangeNotification, target: self, action: #selector(refreshBarFrame))

        if !hideToast {
            app.setRightBarItem(UIBarButtonItem.SystemItem.action) { [weak self] _ in
                self?.app.showSheet(title: nil, message: nil, actions: [self?.globalThread == true ? "切换主线程调用" : "切换后台线程调用", "启用导航栏转场优化"], actionBlock: { [weak self] index in
                    if index == 0 {
                        self?.globalThread.toggle()
                        self?.refreshBarFrame()
                    } else {
                        UINavigationController.app.enableBarTransition()
                    }
                })
            }
        } else {
            app.setLeftBarItem(Icon.closeImage) { [weak self] _ in
                self?.app.close()
            }
        }
    }

    func setupTableStyle() -> UITableView.Style {
        .grouped
    }

    func setupTableLayout() {
        view.addSubview(frameLabel)
        frameLabel.app.layoutChain
            .left(10)
            .right(10)
            .bottom(APP.safeAreaInsets.bottom + 10)

        tableView.backgroundColor = AppTheme.tableColor
        tableView.app.layoutChain
            .horizontal()
            .top()
            .bottom(toViewTop: frameLabel, offset: -10)
    }

    func setupSubviews() {
        tableData.append(contentsOf: [
            ["状态栏切换", "onStatusBar"],
            ["状态栏样式", "onStatusStyle"],
            ["导航栏切换", "onNavigationBar"],
            ["导航栏样式", "onNavigationStyle"],
            ["标题栏颜色", "onTitleColor"],
            ["大标题切换", "onLargeTitle"],
            ["标签栏切换", "onTabBar"],
            ["工具栏切换", "onToolBar"],
            ["导航栏转场", "onTransitionBar"]
        ])
        if !hideToast {
            tableData.append(contentsOf: [
                ["Present(默认)", "onPresent"],
                ["Present(FullScreen)", "onPresent2"],
                ["Present(PageSheet)", "onPresent3"],
                ["Present(默认带导航栏)", "onPresent4"],
                ["Present(Popover)", "onPresent5:"]
            ])
        } else {
            tableData.append(contentsOf: [
                ["Dismiss", "onDismiss"]
            ])
        }
        tableData.append(contentsOf: [
            ["设备转向", "onOrientation"]
        ])
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        if !hideToast {
            UIWindow.app.showMessage(text: "viewWillAppear: \(animated)")
        }
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)

        if !hideToast {
            UIWindow.app.showMessage(text: "viewWillDisappear: \(animated)")
        }
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        refreshBarFrame()
    }

    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        tableData.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = UITableViewCell.app.cell(tableView: tableView)
        let rowData = tableData[indexPath.row]
        cell.textLabel?.text = rowData[0]
        return cell
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let rowData = tableData[indexPath.row]
        _ = perform(NSSelectorFromString(rowData[1]), with: indexPath)
    }

    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        refreshBarFrame()
    }

    @objc func refreshBarFrame() {
        if !globalThread {
            let barText = String(
                format: "全局状态栏：%.0f 当前状态栏：%.0f\n全局导航栏：%.0f 当前导航栏：%.0f\n全局顶部栏：%.0f 当前顶部栏：%.0f\n全局标签栏：%.0f 当前标签栏：%.0f\n全局工具栏：%.0f 当前工具栏：%.0f\n全局安全区域：{%.0f, %.0f, %.0f, %.0f}\n设备横屏：%@ 界面横屏：%@\n全面屏：%@ 灵动岛：%@",
                UIScreen.app.statusBarHeight, app.statusBarHeight,
                UIScreen.app.navigationBarHeight, app.navigationBarHeight,
                UIScreen.app.topBarHeight, app.topBarHeight,
                UIScreen.app.tabBarHeight, app.tabBarHeight,
                UIScreen.app.toolBarHeight, app.toolBarHeight,
                UIScreen.app.safeAreaInsets.top, UIScreen.app.safeAreaInsets.left, UIScreen.app.safeAreaInsets.bottom, UIScreen.app.safeAreaInsets.right,
                String(describing: UIDevice.app.isDeviceLandscape), String(describing: UIScreen.app.isInterfaceLandscape),
                String(describing: UIScreen.app.isNotchedScreen), String(describing: UIScreen.app.isDynamicIsland)
            )
            frameLabel.text = barText
        } else {
            DispatchQueue.global().async { [weak self] in
                let barText = String(
                    format: "设备尺寸：(%.0f, %.0f) 屏幕尺寸：(%.0f, %.0f)\n是否是iPad：%@ 屏幕缩放比：%.0f\n设计图尺寸：(%.0f, %.0f)\n宽度缩放比：%.3f 高度缩放比：%.3f\n设备横屏：%@ 界面横屏：%@\nIDFV: %@\nIDFA: %@\nUUID: %@",
                    UIDevice.app.deviceWidth, UIDevice.app.deviceHeight, UIScreen.app.screenWidth, UIScreen.app.screenHeight,
                    String(describing: UIDevice.app.isIpad), UIScreen.app.screenScale,
                    UIScreen.app.referenceSize.width, UIScreen.app.referenceSize.height,
                    UIScreen.app.relativeScale, UIScreen.app.relativeHeightScale,
                    String(describing: UIDevice.app.isDeviceLandscape), String(describing: UIScreen.app.isInterfaceLandscape),
                    UIDevice.app.deviceIDFV ?? "",
                    UIDevice.app.deviceIDFA,
                    UIDevice.app.deviceUUID
                )
                DispatchQueue.main.async { [weak self] in
                    self?.frameLabel.text = barText
                }
            }
        }
    }

    @objc func onStatusBar() {
        app.statusBarHidden = !app.statusBarHidden
        refreshBarFrame()
    }

    @objc func onStatusStyle() {
        if app.statusBarStyle == .default {
            app.statusBarStyle = .lightContent
        } else {
            app.statusBarStyle = .default
        }
        refreshBarFrame()
    }

    @objc func onNavigationBar() {
        app.navigationBarHidden = !app.navigationBarHidden
        refreshBarFrame()
    }

    @objc func onNavigationStyle() {
        if app.navigationBarStyle == .default {
            app.navigationBarStyle = .white
        } else {
            app.navigationBarStyle = .default
        }
        refreshBarFrame()
    }

    @objc func onTitleColor() {
        navigationController?.navigationBar.app.titleAttributes = navigationController?.navigationBar.app.titleAttributes != nil ? nil : [NSAttributedString.Key.foregroundColor: AppTheme.buttonColor]
        refreshBarFrame()
    }

    @objc func onLargeTitle() {
        navigationController?.navigationBar.prefersLargeTitles = !APP.safeValue(navigationController?.navigationBar.prefersLargeTitles)
        refreshBarFrame()
    }

    @objc func onTabBar() {
        // iOS18如果hidesBottomBarWhenPushed为true时，需调用tabBar.isHidden才能生效
        let hidden = !app.tabBarHidden
        if hidesBottomBarWhenPushed {
            tabBarController?.tabBar.isHidden = hidden
        }
        app.tabBarHidden = hidden
        refreshBarFrame()
    }

    @objc func onToolBar() {
        if app.toolBarHidden {
            let item = UIBarButtonItem.app.item(object: UIBarButtonItem.SystemItem.cancel.rawValue, target: self, action: #selector(onToolBar))
            let item2 = UIBarButtonItem.app.item(object: UIBarButtonItem.SystemItem.done.rawValue, target: self, action: #selector(onPresent))
            toolbarItems = [item, item2]
            app.toolBarHidden = false
        } else {
            app.toolBarHidden = true
        }
        refreshBarFrame()
    }

    @objc func onPresent() {
        let vc = TestAdaptiveController()
        vc.app.presentationDidDismiss = { @MainActor @Sendable in
            UIWindow.app.showMessage(text: "presentationDidDismiss")
        }
        vc.app.completionHandler = { _ in
            UIWindow.app.showMessage(text: "completionHandler")
        }
        vc.hideToast = true
        present(vc, animated: true)
    }

    @objc func onPresent2() {
        let vc = TestAdaptiveController()
        vc.hideToast = true
        vc.modalPresentationStyle = .fullScreen
        present(vc, animated: true)
    }

    @objc func onPresent3() {
        let vc = TestAdaptiveController()
        vc.app.presentationDidDismiss = { @MainActor @Sendable in
            UIWindow.app.showMessage(text: "presentationDidDismiss")
        }
        vc.app.completionHandler = { _ in
            UIWindow.app.showMessage(text: "completionHandler")
        }
        vc.hideToast = true
        vc.modalPresentationStyle = .pageSheet
        present(vc, animated: true)
    }

    @objc func onPresent4() {
        let vc = TestAdaptiveController()
        vc.hideToast = true
        let nav = UINavigationController(rootViewController: vc)
        nav.app.presentationDidDismiss = { @MainActor @Sendable in
            UIWindow.app.showMessage(text: "presentationDidDismiss")
        }
        nav.app.completionHandler = { _ in
            UIWindow.app.showMessage(text: "completionHandler")
        }
        present(nav, animated: true)
    }

    @objc func onPresent5(_ indexPath: IndexPath) {
        if presentedViewController != nil {
            dismiss(animated: true)
            return
        }

        let vc = TestAdaptiveController()
        vc.hideToast = true
        vc.preferredContentSize = CGSize(width: APP.screenWidth / 2, height: APP.screenHeight / 2)
        vc.app.setPopoverPresentation({ [weak self] controller in
            controller.barButtonItem = self?.navigationItem.rightBarButtonItem
            controller.permittedArrowDirections = .up
            let cell = self?.tableView.cellForRow(at: indexPath)
            controller.passthroughViews = cell != nil ? [cell!] : []
        }, shouldDismiss: Bool.random())
        vc.app.presentationDidDismiss = { @MainActor @Sendable in
            UIWindow.app.showMessage(text: "presentationDidDismiss")
        }
        vc.app.completionHandler = { _ in
            UIWindow.app.showMessage(text: "completionHandler")
        }
        present(vc, animated: true)
    }

    @objc func onDismiss() {
        dismiss(animated: true) { [weak self] in
            NSLog("self: %@", String(describing: self))
        }
    }

    @objc func onTransitionBar() {
        navigationController?.pushViewController(TestAdaptiveChildController(), animated: true)
    }

    @objc func onOrientation() {
        if UIDevice.app.isDeviceLandscape {
            UIDevice.app.setDeviceOrientation(.portrait)
        } else {
            UIDevice.app.setDeviceOrientation(.landscapeLeft)
        }
        refreshBarFrame()
    }
}

class TestAdaptiveChildController: UIViewController, ViewControllerProtocol {
    var index: Int = 0

    func setupNavbar() {
        app.extendedLayoutEdge = .all
        if index < 1 {
            app.navigationBarStyle = .default
        } else if index < 2 {
            app.navigationBarStyle = .white
        } else if index < 3 {
            app.navigationBarStyle = .transparent
        } else {
            app.navigationBarStyle = .init([-1, 0, 1, 2].randomElement()!)
            app.navigationBarHidden = app.navigationBarStyle.rawValue == -1
        }
        navigationItem.title = "标题:\(index + 1) 样式:\(app.navigationBarStyle.rawValue)"

        app.setRightBarItem("打开界面") { [weak self] _ in
            let vc = TestAdaptiveChildController()
            vc.index = APP.safeValue(self?.index) + 1
            self?.navigationController?.pushViewController(vc, animated: true)
        }
        view.app.addTapGesture { [weak self] _ in
            let vc = TestAdaptiveChildController()
            vc.index = APP.safeValue(self?.index) + 1
            self?.navigationController?.pushViewController(vc, animated: true)
        }
    }
}
